### PR TITLE
Laplacian filter for the image module

### PR DESCRIPTION
**Pull request by Pulkit-Khandelwal**
The Laplacian can either be implemented by the LoG formula or by placing the coefficients as given in this PR. Both 1D and 2D filter has been implemented with tests.

### CODE REVIEW

**Comment by Pulkit-Khandelwal** [comment]
tests added.

File: torchgeometry/image/laplacian.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        [  1.,   1.,   1.,   1.,   1.],
+        [  1.,   1., -24.,   1.,   1.],
+        [  1.,   1.,   1.,   1.,   1.],
+        [  1.,   1.,   1.,   1.,   1.]])
+
+    """
+    if not isinstance(ksize, int) or ksize % 2 == 0 or ksize <= 0:
+        raise TypeError("ksize must be a tuple of length two. Got {}"
+                        .format(ksize))
+
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: are you sure is this correct?
https://github.com/opencv/opencv/blob/7fb70e170154d064ef12d8fec61c0ae70812ce3d/modules/imgproc/src/deriv.cpp#L797

File: torchgeometry/image/laplacian.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        if not len(x.shape) == 4:
+            raise ValueError("Invalid input shape, we expect BxCxHxW. Got: {}"
+                             .format(x.shape))
+        # prepare kernel
+        b, c, h, w = x.shape
+        tmp_kernel: torch.Tensor = self.kernel.to(x.device).to(x.dtype)
+        kernel: torch.Tensor = tmp_kernel.repeat(c, 1, 1, 1)
+
+        # convolve tensor with gaussian kernel
+        return conv2d(x, kernel, padding=self._padding, stride=1, groups=c)
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: mark as a todo to implement filter2D: https://github.com/opencv/opencv/blob/7fb70e170154d064ef12d8fec61c0ae70812ce3d/modules/imgproc/src/deriv.cpp#L822

File: torchgeometry/image/laplacian.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+def laplacian_blur(src: torch.Tensor,
+                   kernel_size: int) -> torch.Tensor:
+    r"""Function that blurs a tensor using a Laplacian filter.
+
+    The operator smooths the given tensor with a laplacian kernel by convolving
+    it to each channel. It suports batched operation.
+
+    Arguments:
+        src (Tensor): the input tensor.
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: torch.Tensor (otherwise won't link)

File: torchgeometry/image/laplacian.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
+from typing import Tuple
+
+import math
+import torch
+import torch.nn as nn
+from torch.nn.functional import conv2d
+
+
+def laplacian(window_size):
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: remember about typing
```python
def laplacian(window_size: int) -> torch.Tensor:
```

File: torchgeometry/image/laplacian.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    r"""One could also use the Laplacian of Gaussian formula
+        to design the filter.
+    """
+
+    laplacian = torch.ones(window_size)
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: typing: ```laplacian: torch.Tensor = torch.ones(window_size)```

File: torchgeometry/image/laplacian.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+import math
+import torch
+import torch.nn as nn
+from torch.nn.functional import conv2d
+
+
+def laplacian(window_size):
+    r"""One could also use the Laplacian of Gaussian formula
+        to design the filter.
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: why not using laplace formula? we should verify that is compliant with opencv or at least see if can be improved the way to create the kernels

File: torchgeometry/image/laplacian.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    """
+
+    laplacian = torch.ones(window_size)
+    laplacian[int(math.floor(window_size / 2))] = 1 - window_size
+    return laplacian
+
+
+def get_laplacian_kernel(ksize: int):
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: typing needs return type ``` def get_laplacian_kernel(ksize: int) -> torch.Tensor:````

File: torchgeometry/image/laplacian.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    r"""Function that returns the coefficients of a 1D Laplacian filter
+
+    Args:
+        ksize (int): filter size. It should be odd and positive.
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: I would rename ksize to kernel_size

File: test/test_image.py [hunk 9]
```diff
@@ -48,3 +49,27 @@ def test_gradcheck(self):
         input = utils.tensor_to_gradcheck_var(input)  # to var
         assert gradcheck(image.gaussian_blur, (input, kernel_size, sigma,),
                          raise_exception=True)
+
+
+@pytest.mark.parametrize("window_size", [5])
+def test_get_laplacian_kernel(window_size):
```
Comment (edgarriba) [inline test/test_image.py]: add some numeric unittest. Introduce manually input/expected values and use torch.testing assert_allclose

File: test/test_image.py [hunk 10]
```diff
@@ -48,3 +49,27 @@ def test_gradcheck(self):
...
+
+
+class TestLaplacianBlur:
+    @pytest.mark.parametrize("batch_shape",
+                             [(1, 4, 8, 15), (2, 3, 11, 7)])
+    def test_laplacian_blur(self, batch_shape, device_type):
+        kernel_size = 5
+
+        input = torch.rand(batch_shape).to(torch.device(device_type))
+        laplace = image.LaplacianBlur(kernel_size)
```
Comment (edgarriba) [inline test/test_image.py]: add a test and use opencv to compute the laplacian for a given input to see if we match

File: torchgeometry/image/laplacian.py (hunk 2 above)
Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: Yes, indeed! We should also look at OpenCV's separable filter when actually implementing a generic filter function.

File: torchgeometry/image/laplacian.py (hunk 1 above)
Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: It looks like OpenCV supports for kernel sizes: 3,5, and 7  according to:

https://docs.opencv.org/2.4/modules/imgproc/doc/filtering.html?highlight=laplacian#void%20getDerivKernels(OutputArray%20kx,%20OutputArray%20ky,%20int%20dx,%20int%20dy,%20int%20ksize,%20bool%20normalize,%20int%20ktype)

The kernel coefficients you referred to in this link are for a 3 by 3 filter. The formula which I wrote is much more generic one as described here:

https://homepages.inf.ed.ac.uk/rbf/HIPR2/log.htm

File: torchgeometry/image/laplacian.py (hunk 6 above)
Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: see answer before. OpenCV has hardcoded for kernel_size of 3 and for the rest I think it uses Sobel filter coefficients (from what I follow). Please, correct me if I'm wrong.

File: torchgeometry/image/laplacian.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        >>> input = torch.rand(2, 4, 5, 5)
+        >>> laplace = tgm.image.LaplacianBlur(5)
+        >>> output = laplace(input)  # 2x4x5x5
+    """
+
+    def __init__(self, kernel_size: int) -> None:
+        super(LaplacianBlur, self).__init__()
+        self.kernel_size: int = kernel_size
+        self._padding: int = self.compute_zero_padding(kernel_size)
+        self.kernel: torch.Tensor = self.get_laplacian_kernel(kernel_size)
```
[the remark this answers is not in the document]
Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: kept it there in spirit of the Gaussian function. What about the edge pixels?

File: test/test_image.py [hunk 12]
```diff
@@ -1,8 +1,10 @@
 import pytest
 
 import torch
+import math
 import torchgeometry.image as image
 from torch.autograd import gradcheck
+import torch_testing as tt
```
Comment (edgarriba) [inline test/test_image.py]: ```from torch.testing import assert_allclose```

File: test/test_image.py [hunk 13]
```diff
@@ -1,8 +1,10 @@
 import pytest
 
 import torch
+import math
```
Comment (edgarriba) [inline test/test_image.py]: use torch random generators

File: docker/Dockerfile [hunk 14]
```diff
@@ -15,4 +15,5 @@ RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-la
      /opt/conda/bin/conda clean -ya
 ENV PATH /opt/conda/bin:$PATH
 
+RUN pip3 install torch_testing
```
Comment (edgarriba) [inline docker/Dockerfile]: don't need that

File: test/test_image.py [hunk 15]
```diff
@@ -48,3 +50,34 @@ def test_gradcheck(self):
...
+
+
+@pytest.mark.parametrize("window_size", [5])
+def test_get_laplacian_kernel(window_size):
+    kernel = image.get_laplacian_kernel(window_size)
+    assert kernel.shape == (window_size,)
+    assert kernel.sum().item() == pytest.approx(0.0)
+
+
+@pytest.mark.parametrize("window_size", [5])
```
Comment (edgarriba) [inline test/test_image.py]: if you don't test multiple values, just remove it

File: torchgeometry/image/laplacian.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply torchgeometry/image/laplacian.py]: in gaussian could be removed too. Which edge pixels?

File: torchgeometry/image/laplacian.py (hunk 6 above)
Comment (edgarriba) [reply torchgeometry/image/laplacian.py]: ok, let me verify it

File: torchgeometry/image/laplacian.py [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        [  1.,   1.,   1.,   1.,   1.]])
+
+    """
+    if not isinstance(kernel_size, int) or kernel_size % 2 == 0 or \
+            kernel_size <= 0:
+        raise TypeError("ksize must be an odd positive integer. Got {}"
+                        .format(kernel_size))
+
+    kernel = torch.ones((kernel_size, kernel_size))
+    mid = int(math.floor((kernel_size / 2)))
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: torch.floor, torch.pow

File: test/test_image.py (hunk 9 above)
Comment (edgarriba) [reply test/test_image.py]: I was meaning to have something like
```python
class TestGetLaplacianFilter:
    def test_get_filter_w3(...):
        pass
    def test_get_filter_w5(...):
        pass
```
consider add as needed

File: test/test_image.py [hunk 17]
```diff
@@ -48,3 +49,36 @@ def test_gradcheck(self):
...
+
+class TestLaplacian:
+    @pytest.mark.parametrize("batch_shape",
+                             [(1, 4, 8, 15), (2, 3, 11, 7)])
+    def test_laplacian(self, batch_shape, device_type):
+        kernel_size = 5
+
+        input = torch.rand(batch_shape).to(torch.device(device_type))
+        laplace = image.Laplacian(kernel_size)
+        assert laplace(input).shape == batch_shape
```
Comment (edgarriba) [inline test/test_image.py]: You might also need a gradient check function. Additionally, try to play with jitted module. The plan is also to support but it's something that I haven't tried at all

File: torchgeometry/image/laplacian.py (hunk 16 above)
Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: These functions apply to tensors. What I want is just an int.

Comment (Pulkit-Khandelwal) [reply torchgeometry/image/laplacian.py]: Though I can convert it to introduce. So would change

Comment (edgarriba) [reply torchgeometry/image/laplacian.py]: oh wait. just do: kernel_size // 2

File: test/test_image.py (hunk 10 above)
Comment (Pulkit-Khandelwal) [reply test/test_image.py]: Let me know after you verify which implementation of Laplacian do we want? Then, I can add this test.

**Comment by Pulkit-Khandelwal** [comment]
Also, I wonder whether all these filters such as laplacian, gaussain, sobel and more to come to under one module? like filters.py rather than gaussain.py, laplacian.py etc.? And, we *could* have a separate submodule for the functional interface?

File: torchgeometry/image/laplacian.py [hunk 18]
```diff
@@ -0,0 +1,175 @@
+from typing import Tuple
+
+import math
```
Comment (edgarriba) [inline torchgeometry/image/laplacian.py]: remove this if you don't use it anymore

File: test/test_image.py (hunk 10 above)
Comment (edgarriba) [reply test/test_image.py]: I don't see why opencv uses sobel filter for laplacian, but after a bit of research I think that in the way we are doing should be fine. @bhack do you have any insight why is that ?

I've found this interesting test: https://stackoverflow.com/questions/32768407/what-is-the-laplacian-mask-kernel-used-in-the-scipy-ndimage-filter-laplace

**Comment by edgarriba** [comment]
@Pulkit-Khandelwal implement the missing tests: unit and gradient check as well. Regarding the organization, I would like to keep it like this and once we have a more robust api we can figure out how reorganize all the modules. I actually like it in that way because all the codes are self-contained.

File: test/test_image.py (hunk 10 above)
Comment (Pulkit-Khandelwal) [reply test/test_image.py]: I did some more digging.
Refer the link:
https://github.com/scipy/scipy/blob/v0.16.0/scipy/ndimage/filters.py#L396

Please, run the following code snippet (to verify) which I wrote to see what scipy spits out:
``
from scipy.ndimage import *
import numpy as np

A = np.zeros((7,7))
A[3,3] = 1
print(A)

#I have extended the filter code in scipy to account for any kernel size [1, 1, -4, 1, 1] kernel size=5 here:

def derivative2(input, axis, output, mode, cval):
        return correlate1d(input, [1, 1, -4, 1, 1], axis, output, mode, cval, 0)

B = generic_laplace(A, derivative2)
print(B)

#And this is what scipy gives because it uses [1, -2, 1] as the only kernel size of 3
C= scipy.ndimage.filters.laplace(A)

print(C)
``
A = array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

B = array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  1.,  1., -8.,  1.,  1.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.]])

C = array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  1., -4.,  1.,  0.,  0.],
       [ 0.,  0.,  0.,  1.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.]])

Now, one would ask is why B slightly different from what I have in PyTorch geometry?
Because, I consider cross-derivatives as well. See here:

D =  [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., -24., 1, 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],

File: test/test_image.py (hunk 9 above)
Comment (Pulkit-Khandelwal) [reply test/test_image.py]: This looks neat. What about the similar funcitons in tests for the gaussian. SHoudl they change too? Right now it is in the spirit similar to that.

File: test/test_image.py (hunk 10 above)
Comment (Pulkit-Khandelwal) [reply test/test_image.py]: Hope this now clears the differences between scipy, OpenCV and my implementation.

File: test/test_image.py (hunk 9 above)
Comment (Pulkit-Khandelwal) [reply test/test_image.py]: I'll explore more tests functions available in PyTorch in due time.

**Comment by edgarriba** [comment]
closed in favour of https://github.com/arraiyopensource/torchgeometry/pull/118
